substrate/client/rpc-servers/src/middleware/chain_head.rs

```diff
@@ -0,0 +1,237 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! RPC middleware to ensure chainHead methods are called from a single connection.
+
+use std::{
+	collections::HashSet,
+	future::Future,
+	pin::Pin,
+	sync::Arc,
+	task::{Context, Poll},
+};
+
+use jsonrpsee::{
+	server::middleware::rpc::RpcServiceT,
+	types::{Params, Request},
+	MethodResponse,
+};
+use parking_lot::Mutex;
+use pin_project::pin_project;
+
+/// The per connectin data needed to manage chainHead subscriptions.
+#[derive(Default)]
+pub struct ConnectionData {
+	/// Active `chainHead_follow` subscriptions for this connection.
+	subscriptions: HashSet<String>,
+}
+
+/// Layer to allow the `chainHead` RPC methods to be called from a single connection.
+#[derive(Clone)]
+pub struct ChainHeadLayer {
+	connection_data: Arc<Mutex<ConnectionData>>,
+}
+
+impl ChainHeadLayer {
+	/// Create a new [`ChainHeadLayer`].
+	pub fn new(connection_data: Arc<Mutex<ConnectionData>>) -> Self {
+		Self { connection_data }
+	}
+}
+
+impl<S> tower::Layer<S> for ChainHeadLayer {
+	type Service = ChainHeadMiddleware<S>;
+
+	fn layer(&self, inner: S) -> Self::Service {
+		ChainHeadMiddleware::new(inner, self.connection_data.clone())
+	}
+}
+
+/// Chain head middleware.
+#[derive(Clone)]
+pub struct ChainHeadMiddleware<S> {
+	service: S,
+	connection_data: Arc<Mutex<ConnectionData>>,
+}
+
+impl<S> ChainHeadMiddleware<S> {
+	/// Create a new chain head middleware.
+	pub fn new(service: S, connection_data: Arc<Mutex<ConnectionData>>) -> ChainHeadMiddleware<S> {
+		ChainHeadMiddleware { service, connection_data }
+	}
+}
+
+impl<'a, S> RpcServiceT<'a> for ChainHeadMiddleware<S>
+where
+	S: Send + Sync + RpcServiceT<'a>,
+{
+	type Future = ResponseFuture<S::Future>;
+
+	fn call(&self, req: Request<'a>) -> Self::Future {
+		const CHAIN_HEAD_FOLLOW: &str = "chainHead_unstable_follow";
+		const CHAIN_HEAD_CALL_METHODS: [&str; 8] = [
```

Comment:
A random thought is whether we can fetch these method names from the rpc module itself i.e, to keep in sync if a method is removed, modified and so on.

Perhaps, we could add some integration tests for it at some point.